service/storage_service.cc

```diff
@@ -6038,6 +6077,187 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
     }
 }
 
+future<join_node_request_result> storage_service::join_node_request_handler(join_node_request_params params) {
+    if (!_raft_topology_change_enabled) {
+        throw std::runtime_error("The node is not configured to use topology over raft");
+    }
+
+    join_node_request_result result;
+    slogger.info("raft topology: received request to join from host_id: {}, join cookie: {}", params.host_id, params.join_cookie);
+
+    if (params.cluster_name != _db.local().get_config().cluster_name()) {
+        result.result = join_node_request_result::rejected{
+            .reason = ::format("Cluster name check failed. This node cannot join the cluster "
+                                "because it expected cluster name \"{}\" and not \"{}\"",
+                                params.cluster_name,
+                                _db.local().get_config().cluster_name()),
+        };
+        co_return result;
+    }
+
+    if (params.snitch_name != _db.local().get_snitch_name()) {
+        result.result = join_node_request_result::rejected{
+            .reason = ::format("Snitch name check failed. This node cannot join the cluster "
+                                "because it uses \"{}\" and not \"{}\"",
+                                params.snitch_name,
+                                _db.local().get_snitch_name()),
+        };
+        co_return result;
+    }
+
+    while (true) {
+        auto guard = co_await _group0->client().start_operation(&_abort_source);
+
+        if (_topology_state_machine._topology.is_empty()) {
+            // We need to wait until the discovery leader (us?) puts its join
+            // request. Tell the node to try again later.
+            result.result = join_node_request_result::in_progress {};
+            co_return result;
+        }
+
+        if (_topology_state_machine._topology.left_nodes.contains(params.host_id)) {
+            result.result = join_node_request_result::rejected{
+                .reason = "The node has already been removed from the cluster",
+            };
+            co_return result;
+        }
+
+        if (params.replaced_id.has_value()) {
+            auto it = _topology_state_machine._topology.normal_nodes.find(*params.replaced_id);
+            if (it == _topology_state_machine._topology.normal_nodes.end()) {
+                result.result = join_node_request_result::rejected {
+                    .reason = ::format("Cannot replace node {} because it is not in the 'normal' state", *params.replaced_id),
+                };
```

Comment:
This is pre-existing but the check should be in the coordinator before is starts executing the command since by then the state of the node may change.

Reply:
Forgot to fix this before, but now this check is moved to the topology coordinator (`validate_joining_node`).